aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/EventSourceType.java

```diff
@@ -0,0 +1,194 @@
+     * from the logic that handles the notfication
+     * @return
+     */
+    public interface EventSourceAttributeParser {
+        Map<String, String> getEventSourceAttributes(Bundle bundle);
+    }
+
+    private static final EventSourceAttributeParser CAMPAIGN_PARSER = new EventSourceAttributeParser() {
+        @Override
+        public Map<String, String> getEventSourceAttributes(Bundle bundle) {
+            if(bundle == null) {
+                return null;
+            }
+            final Map<String, String> campaignAttributes = new HashMap<>();
+            String campaignKeyPrefix = CAMPAIGN.getEventSourceKeyPrefix();
+
+            for(String key : bundle.keySet()) {
+                if (key.startsWith(campaignKeyPrefix)) {
+                    //If it matches the prefix, include it in the attributes
+                    //after stripping out the prefix
+                    campaignAttributes.put(key.replace(campaignKeyPrefix,""), bundle.getString(key));
+                }
+            }
+            return campaignAttributes.size() > 0 ? campaignAttributes : null;
+        }
+    };
+
+    private static final EventSourceAttributeParser JOURNEY_PARSER = new EventSourceAttributeParser() {
+        @Override
+        public Map<String, String> getEventSourceAttributes(Bundle bundle) {
+            String pinpointPayload = bundle.getString(PINPOINT_ATTRIBUTE_KEY);
+            if(pinpointPayload == null) {
+                return null;
+            }
+            Map<String, String> result = null;
+            try {
+                JsonElement pinpointJson = new JsonParser().parse(pinpointPayload);
+                JsonElement journeyJson = pinpointJson.getAsJsonObject().get(JOURNEY_ATTRIBUTE_KEY);
+                if(journeyJson == null) {
+                    return null;
+                }
+                result = new HashMap<>();
+                for(Map.Entry entry : journeyJson.getAsJsonObject().entrySet()) {
+                    result.put((String) entry.getKey(), ((JsonPrimitive)entry.getValue()).getAsString());
+                }
+            } catch(JsonSyntaxException ex) {
+                Log.w(TAG, "Exception attempting to parse pinpoint JSON payload.", ex);
+                Log.v(TAG, "Payload: "+pinpointPayload);
+                return null;
+            }
+
+            return result;
+        }
+    };
+
+    public static EventSourceType CAMPAIGN = new EventSourceType("campaign",
+            "_campaign",
+            CAMPAIGN_ID_ATTRIBUTE_KEY,
+            CAMPAIGN_ACTIVITY_ID_ATTRIBUTE_KEY,
+            CAMPAIGN_PUSH_KEY_PREFIX,
+            CAMPAIGN_PARSER);
+    public static EventSourceType JOURNEY = new EventSourceType("journey",
+            "_journey",
+            JOURNEY_ID_ATTRIBUTE_KEY,
+            JOURNEY_ACTIVITY_ID_ATTRIBUTE_KEY,
+            null,
+            JOURNEY_PARSER);
+
+    public static EventSourceType getEventSourceType(Bundle bundle) {
```

Comment:
Can you audit all of these methods -- how many need to be public? Can any be private or default (pakage-local) access?

Reply:
I've gone through and there are only a couple things that remain public now.